src/lib.rs

```diff
@@ -54,347 +58,456 @@ impl Default for Duration {
     }
 }
 
-#[derive(Serialize, Debug, Builder)]
-#[builder(setter(into))]
-#[serde(rename_all = "camelCase")]
-pub struct ChartParams {
-    #[builder(default)]
-    pub chart_reset: Option<bool>,
-    #[builder(default)]
-    pub chart_simplify: Option<bool>,
-    #[builder(default)]
-    pub chart_interval: Option<i64>,
+#[derive(PartialEq, Eq)]
+/// The `Request` enum type allows for HTTP request matching to the IEX API.
+// TODO(Request): Add documentation from IEX website.
+pub enum Request<'a> {
+    Book {
+        symbol: &'a str,
+    },
+    // REVIEW(CHART): Should `time-series` be implemented even though it forwards
+    // towards the same endpoint as `Chart`?
+    Chart {
+        symbol: &'a str,
+        duration: Duration,
+        params: Option<Vec<ChartParam>>,
+    },
+    Company {
+        symbol: &'a str,
+    },
+    DelayedQuote {
+        symbol: &'a str,
+    },
+    Dividends {
+        symbol: &'a str,
+        duration: Duration,
+    },
+    Earnings {
+        symbol: &'a str,
+    },
+    EffectiveSpread {
+        symbol: &'a str,
+    },
+    Financials {
+        symbol: &'a str,
+    },
+    // REVIEW|TODO(List): add in default displayParameters(?)
+    List {
+        param: ListParam,
+    },
+    Logo {
+        symbol: &'a str,
+    },
+    News {
+        symbol: &'a str,
+        range: Option<i32>,
+    },
+    Ohlc {
+        symbol: &'a str,
+    },
+    Peers {
+        symbol: &'a str,
+    },
+    Previous {
+        symbol: &'a str,
+    },
+    Price {
+        symbol: &'a str,
+    },
+    Quote {
+        symbol: &'a str,
+    },
+    Relevant {
+        symbol: &'a str,
+    },
+    Splits {
+        symbol: &'a str,
+        duration: Duration,
+    },
+    Stats {
+        symbol: &'a str,
+    },
+    Symbols,
+    /// IEX Regulation SHO Threshold Securities List
+    ThresholdSecurities {
+        // REVIEW: may be a good idea to implement a date struct to use here and
+        // wrap in the Duration::Date enum variant, or eq - a match statement.
+        date: Option<Duration>,
+    },
+    // TODO(ShortInterest): implement variant.
+    VolumeByVenue {
+        symbol: &'a str,
+    },
 }
 
-impl Default for ChartParams {
-    fn default() -> ChartParams {
-        ChartParamsBuilder::default().build().unwrap()
+impl<'a> ToString for Request<'a> {
```

Comment:
I'm not 100% on this implementation, because it makes you jump around to figure out how URIs are formed.

I preferred the approach of having a method per API endpoint and putting all of the logic in each of those methods. What benefit does this proposed new approach have over the old one?

Reply:
@samwho I'm not sure that I am on the same page as you are. If you are referring specifically to the building of the implementation of the chart params url then I can easily take away the parse params function and insert the logic into the ChartParms block of the Request ToString method. Regarding specifically why I did this, is because at the time it made sense to follow the logic of the Duration enum. Also, the method does not need to be exposed to the public. If I did not address your question, please let me know. 

---

Regarding the benefits of just one API endpoint: 
- We are exposing less to the end-user - (benefit on its own)
-  The multiple endpoints method signatures had the same behaviour, which was to request a Response - whether that would be dividends, earnings, etc. I do not think that it is idiomatic to expose intentions of state for a method, instead of behaviour. With the multiple-endpoint API, we expose state, and hide behaviour. Meaning that dividends (inferring state) is a `pub fn`, but `get` which infers behaviour, is hidden. 
    - However, with a single endpoint the method signatures are the same. The multiple endpoints repeat the same behaviour with different inputs (state variables), and all we are doing with the request enum is coalescing all of the multiple endpoint's states into one type, which all have the same behaviour. 
- This helps maintainability because intentions are inherently as well as easily expressed.
- Method signatures are the same. We return a serde_json::Value instead of a custom type. Since there is no behaviour wrapped around the structs found in types.rs, there is no reason to return the value as the specific type when it has the same functionality and behavior as serde_json::Value. We leave it up the end-user if they want to change the return struct's type. 
-  Having the type-variant-enum approach provides bounds for maintainability. Instead of creating another method, we just add to our request enum, and update the ToString method
    -  Have you thought about how you are going to implement batch requests? Allowing a variadic Request argument to this particular method is going to make implementation extremely easy. Personally, I am not sure how batch requests would be implemented with multiple endpoints. First off, what would be the input arguments? Nested methods or a variadic method signature is not exactly elegant. Personally, I would suggest a variadic argument consisting of Request enum type variants. This way, we can worry about state, and expose behaviour. However, if we stick with multiple endpoints, you can either use serde_json::Value as a return type wrapped in a result, or, we would need to create 2^n method signatures for each mix and match of the methods - which I think you would agree, should not be used for batch requests.

I think this is enough to express the majority of my ideas for the single-point implementation, but if need be, I can provide more input.

Reply:
That makes sense.

I like to try and make the surface of APIs that are mostly wrapping calls to HTTP endpoints look like those endpoints as much as possible, but agree that this approach has merit and am happy to go forward with it.